presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1224,15 +1215,23 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
                     if (!typeManager.canCoerce(type, superType)) {
                         throw new SemanticException(TYPE_MISMATCH, expression, message, superType);
                     }
-                    expressionCoercions.put(expression, superType);
-                    if (typeManager.isTypeOnlyCoercion(type, superType)) {
-                        typeOnlyCoercions.add(expression);
-                    }
+                    addOrDeleteExpression(expression, type, superType);
                 }
             }
 
             return superType;
         }
+
+        private void addOrDeleteExpression(Expression expression, Type type, Type superType)
```

Comment:
This name doesn't seem right.
First of all it's not `addOrDelete` rather `addOrReplace` (the fact that it can remove expression from typeOnlyCoercions is only detail of data representation).

Also, I would rather call it `addOrReplaceExpressionCoercion` or `addOrReplaceCoercion`, as this is semantically what we do. Calling that `addOrReplaceExpression` doesn't give to much information about context that this method requires.